### PR TITLE
update to Qt6 on windows, support Qt6 elsewhere

### PR DESCRIPTION
**Pull request by shuffle2**
The main point of this PR is to generate builds that can be tested, track bugs and comments on the [qt submodule update](https://github.com/shuffle2/ext-win-qt/tree/qt6.3.0). Would love review of the [build config](https://github.com/shuffle2/ext-win-qt/blob/qt6.3.0/Qt6.3.0/dolphin-x64.yml). I was able to disable a significant enough portion of the Qt modules such that I could add the compressed pdbs into the repo and still come in at less total size than the previous commit.

I've removed the pdb archives from the qt repo to minimize size as much as possible, but not sure what the best way to keep them available for people is. any ideas?

An ARM64 build is [here](https://drive.google.com/file/d/1vKfsEoC7WPPFwv0IjE-axxk7BY3XVlZq/view?usp=sharing)

Some known issues:
None!, Find some!

Fixed issues:
1. Either Qt or compiler bug relating to atomic ptr template https://bugreports.qt.io/browse/QTBUG-102246 . It seems to be a msvc arm64 compiler bug: https://developercommunity.visualstudio.com/t/codegen-elides-initializers-when-copying/10004323 . (or, can someone point out how the source code is invalid?)
   * Fix Version/s: 6.2.5, 6.3.1 (planned: 13.5.2022), 6.4.0 Beta1
   * I've manually applied the patch for the Qt build. But it only makes a difference for ARM64, anyway.
2. Apparently high dpi scaling issue? https://github.com/dolphin-emu/dolphin/pull/10546#issuecomment-1086663371 Would be great if someone more familiar with Qt could comment on if that's expected / new behavior / what to do / etc.
   * at least initially, I've made Qt6 use Qt5's scaling behavior to fix this
3. There is some change-detection issue with the current msbuild moc setup, which will cause problems for the buildbot if it tries to switch back and forth between Qt5 and Qt6 builds.
   * this should be fixed cleanly (i've changed the build to output the moc files to different directory)

note: this will need the changes in my branch of ext-win-qt submodule merged into the dolphin copy of that repo, and the submodule in this PR switched back.

I've created a PR for the submodule here: https://github.com/dolphin-emu/ext-win-qt/pull/18

### CODE REVIEW

**Comment by JosJuice** [comment]
This PR seems to be causing errors for other PRs: #10560

**Comment by shuffle2** [comment]
> This PR seems to be causing errors for other PRs: #10560

The buildbot should be fixed now, for as long as *this* PR isn't rebuilt. Might have to figure out a real solution in the future if we want the buildbots to keep building PRs based on pre-qt6 and post-qt6 at the same time (?)

**Comment by shuffle2** [comment]
possible resource for adapting Qt6 DPI scaling stuff: https://doc.qt.io/qt-6/highdpi.html (haven't read / dug into in-depth yet)
@dreamsyntax maybe you can try fiddling with some of the mentioned environment variables and see if it changes the effect you mentioned?

File: Source/Core/DolphinQt/Main.cpp [hunk 1]
```diff
@@ -132,8 +132,10 @@ int main(int argc, char* argv[])
   const optparse::Values& options = CommandLineParse::ParseArguments(parser.get(), argc, argv);
   const std::vector<std::string> args = parser->args();
 
+#if (QT_VERSION < QT_VERSION_CHECK(6, 0, 0))
   QCoreApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
```
Comment (shuffle2) [inline Source/Core/DolphinQt/Main.cpp:136]: hm, these 2 flags were deprecated in Qt6, but not clear to me if that implicitly makes some of the existing dolphin code behave badly. AA_UseHighDpiPixmaps was documented as:

> Make [QIcon::pixmap](https://doc.qt.io/qt-5/qicon.html#pixmap)() generate high-dpi pixmaps that can be larger than the requested size. Such pixmaps will have [devicePixelRatio()](https://doc.qt.io/qt-5/qpixmap.html#devicePixelRatio) set to a value higher than 1. After setting this attribute, application code that uses pixmap sizes in layout geometry calculations should typically divide by [devicePixelRatio()](https://doc.qt.io/qt-5/qpixmap.html#devicePixelRatio) to get device-independent layout geometry.

Comment (shuffle2) [reply Source/Core/DolphinQt/Main.cpp:136]: should probably also check if the dpi setting in manifest should be changed/removed https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/DolphinQt/DolphinQt.manifest#L5 (Qt6 is takes care of it now)

**Comment by dreamsyntax** [comment]
> possible resource for adapting Qt6 DPI scaling stuff: https://doc.qt.io/qt-6/highdpi.html (haven't read / dug into in-depth yet) @dreamsyntax maybe you can try fiddling with some of the mentioned environment variables and see if it changes the effect you mentioned?

I'll give it a shot. I assume basing off this branch should be sufficient enough to build?

**Comment by shuffle2** [comment]
> > possible resource for adapting Qt6 DPI scaling stuff: https://doc.qt.io/qt-6/highdpi.html (haven't read / dug into in-depth yet) @dreamsyntax maybe you can try fiddling with some of the mentioned environment variables and see if it changes the effect you mentioned?
> 
> I'll give it a shot. I assume basing off this branch should be sufficient enough to build?

It should be possible to use the build from this pr and set the env vars before launching dolphin. That way you don’t need to build anything yourself. Otherwise yes, you can build this pr.

from a terminal like cmd.exe:
```
set VAR=VAL
Dolphin.exe
```

**Comment by dreamsyntax** [comment]
> set QT_SCALE_FACTOR_ROUNDING_POLICY=Round

Yep, this accomplishes the old behavior
scaling

In case anyone else tries this, only cmd worked in tests.

QT_ENABLE_HIGHDPI_SCALING=0 also looked about the same (though it is not supposed to be used)

If you set the rounding policy in the build we should be good.

**Comment by dreamsyntax** [comment]
> Just wanted to confirm the Qt6 builds of Dolphin no longer have the bug "UI progress bar % freezing unless the mouse is moving over the window"
>
>Including for scenarios such as:
>
 >   Verify
  >  Extract Disc

This issue I posted above is actually the same issue as the save/load hotkey issue, hence why it is resolved.
Additional fixed issue: regression in Qt5.15.1 causing UI freezing will be resolved

This PR below has details and will be closed as upgrading to Qt6 will resolve it.
https://github.com/dolphin-emu/dolphin/pull/10546

**Comment by shuffle2** [comment]
It's ready now with Qt 6.3.0. May 13th is expected Qt 6.3.1 release date, but AFAIK we have no reason to wait for it.
My most recent force-push of this PR was to take care of RTTI issue (see https://github.com/dolphin-emu/ext-win-qt/pull/18#issuecomment-1108809351)

**Comment by mbc07** [comment]
Just gave this PR a try and Dolphin's GUI finally renders at sane sizes on my laptop (2560x1440 resolution on a 15.6" screen, with 150% scaling set on Windows' display settings). On master, everything is rendered at abnormally huge sizes, to the point none of the config windows fit on screen and nearly all text seems to be using bold fonts. This PR fixes all of that! Here's a comparison (master on top, this PR at bottom):

master

this PR

**Comment by dreamsyntax** [comment]
> Just gave this PR a try and Dolphin's GUI finally renders at sane sizes on my laptop (2560x1440 resolution on a 15.6" screen, with 150% scaling set on Windows' display settings). On master, everything is rendered at abnormally huge sizes, to the point none of the config windows fit on screen and nearly all text seems to be using bold fonts. This PR fixes all of that! Here's a comparison (master on top, this PR at bottom):
> 
> master
> 
> this PR

So... it sounds like you are liking the default behavior opposed to running with

`set QT_SCALE_FACTOR_ROUNDING_POLICY=Round`

But this causes some elements to look stretched/off.

Can you send a screenshot of how the ImgUI (Savestate/FPS etc in-game UI) looks with the flag and without?

**Comment by shuffle2** [comment]
@dreamsyntax  fwiw this PR currently uses `Round` (changed since you tested)

**Comment by dreamsyntax** [comment]
Ah cool, assuming @mbc07 built your latest build rather than the build you provided (judging by screenshot I think they did pull latest) then there's no issue :)

**Comment by dreamsyntax** [comment]
> It's ready now with Qt 6.3.0. May 13th is expected Qt 6.3.1 release date, but AFAIK we have no reason to wait for it. My most recent force-push of this PR was to take care of RTTI issue (see [dolphin-emu/ext-win-qt#18 (comment)](https://github.com/dolphin-emu/ext-win-qt/pull/18#issuecomment-1108809351))

Imo may be worth waiting the few extra days for May 13, a new minor version of Qt usually has a bigger patch1.

**Comment by dreamsyntax** [comment]
Actually, I didn't find anything out of the ordinary with the build so maybe we're fine to go with 6.3.0. If there's another issue like the 5.15.1 bug, then we can update at some point.

I'm for going with 6.3.0 asap.

**Comment by mbc07** [comment]
@dreamsyntax the ImGui elements look stretched on master (together with the abnormally huge interface). On this PR they look much more "correct" to me, properly sized in comparison to other programs and to the Windows interface and with no apparent stretching on the ImGui elements from the rendering window...

**Comment by shuffle2** [comment]
I tested that it compiles and runs fine on ubuntu jammy, packages `qt6-base-dev` and `qt6-base-private-dev` are min requirements.

Since it's after 1st of the month, merge now? :)

**Comment by dreamsyntax** [comment]
> I tested that it compiles and runs fine on ubuntu jammy, packages `qt6-base-dev` and `qt6-base-private-dev` are min requirements.
> 
> Since it's after 1st of the month, merge now? :)

Just tried your latest build.
All the issue scenarios seem to be fixed, including the progress bar freezing issue.
ext

File: Source/Core/DolphinQt/CMakeLists.txt [hunk 2]
```diff
@@ -1,15 +1,24 @@
-if (NOT Qt5_DIR AND MSVC)
+if (Qt6_DIR)
+	set(Qt_DIR "${Qt6_DIR}")
+endif()
+
+if (Qt5_DIR)
+	set(Qt_DIR "${Qt5_DIR}")
+endif()
```
Comment (Dentomologist) [inline Source/Core/DolphinQt/CMakeLists.txt:7]: I could be failing to read CMake properly, but it seems like if Qt5_DIR and Qt6_DIR are both set it'll use Qt5. Is that intended?

Comment (shuffle2) [reply Source/Core/DolphinQt/CMakeLists.txt:7]: i think this section is for manually overriding the Qt installation directory, so you assume the user won't call cmake with both directories defined. The auto-finding of the Qt installation should be done by the following `find_package(Qt NAMES Qt6 Qt5...` call, I assume that chooses higher version as expected but haven't tested.

Comment (shuffle2) [reply Source/Core/DolphinQt/CMakeLists.txt:7]: so, you're correct in that all this version detection logic was not working correctly. i'll push what i got working in a few. strangely the documented method doesn't seem to actually work.

**Comment by shuffle2** [comment]
requesting the dolphin gods to push the button

**Comment by JMC47** [comment]
I think we have the Progress Report build marked, so let's see what happens with this.

**Comment by Avasam** [comment]
Fixes an issue with some dropdowns: https://bugs.dolphin-emu.org/issues/12904

**Comment by AhayriSG** [comment]
image

**Comment by refractionpcsx2** [comment]
Dunno if you guys are aware yet (though the last post might have been a clue), but Qt 6.3.0 is only compatible with Windows 10.  Windows 8.1 and below just throw an error, you have to drop back to 6.2 in order for it to work on the older OS's (even though they aren't officially supported by Qt) :)

**Comment by shuffle2** [comment]
We are aware; we are no longer 'supporting' win7/8. For the foreseeable future dolphin should still technically work on those OS (as core dolphin does not itself break compat, people could compile against older Qt themselves or use nogui executable), but I think it's only a matter of time before the legacy compat code in core is removed or breaks creep in.

**Pull request merged**